trunner/runners/wrappers.py

```diff
@@ -0,0 +1,320 @@
+        if is_github_actions():
+            logging.info('::group::Run psu\n')
+
+        while True:
+            line = self.proc.readline()
+            if not line:
+                break
+
+            logging.info(line)
+
+        if is_github_actions():
+            logging.info('::endgroup::\n')
+
+    def run(self):
+        # Use pexpect.spawn to run a process as PTY, so it will flush on a new line
+        self.proc = pexpect.spawn(
+            self.progname,
+            self.args,
+            cwd=self.cwd,
+            encoding='ascii'
+        )
+
+        self.read_output()
+        self.proc.wait()
+        if self.proc.exitstatus != 0:
+            logging.error('psu failed!\n')
+            raise Exception(' Loading plo using psu failed!')
+
+
+class Gdb(StandardWrapper):
+    """Wrapper for gdb-multiarch program"""
+
+    def __init__(self, target, file, script, cwd=None):
+        args = [f'{file}', '-x', f'{script}']
+        super().__init__(target, 'gdb-multiarch', args, cwd)
+
+    def read_output(self):
+        if is_github_actions():
+            logging.info('::group::Run gdb\n')
+
+        self.proc.expect_exact("(gdb) ")
+
+        # TODO: investigate why send('c\r\n') does not work
+        self.proc.send('c')
+        self.proc.send('\r\n')
+
+        logging.info(f'{self.proc.before}\n')
+        self.proc.close()
+
+        if is_github_actions():
+            logging.info('::endgroup::\n')
+
+    def run(self):
+        with GdbServer():
+            # Use pexpect.spawn to run a process as PTY, so it will flush on a new line
+            self.proc = pexpect.spawn(
+                self.progname,
+                self.args,
+                cwd=self.cwd,
+                encoding='ascii'
+            )
+
+            self.read_output()
+            self.proc.close()
+
+    def close(self):
+        self.proc.close()
+        if self.proc.exitstatus != 0:
+            logging.error('gdb failed!\n')
+            raise Exception('Loading plo using gdb failed!')
+
+
+def phd_error_msg(message, output):
+    msg = message
+    msg += Color.colorify('\nPHOENIXD OUTPUT:\n', Color.BOLD)
+    msg += output
+
+    return msg
+
+
+def phd_warning_msg(message, warning):
+    msg = message
+    msg += Color.colorify('\nPHOENIXD WARNING:\n', Color.BOLD)
+    msg += warning
+
+    return msg
+
+
+class PhoenixdError(Exception):
+    pass
+
+
+class Phoenixd(BackgroundWrapper):
+    """ Wrapper for phoenixd program"""
+
+    def __init__(
+        self,
+        target,
+        port,
+        dir='.',
+        cwd=None,
+        wait_dispatcher=True,
+        long_flashing=False
+    ):
+        if cwd is None:
+            cwd = boot_dir(target)
+        self.cwd = cwd
+        self.port = port
+        self.dir = dir
+        self.wait_dispatcher = wait_dispatcher
+        self.dispatcher_event = None
+        self.long_flashing = long_flashing
+        super().__init__()
+
+    def _reader(self):
```

Comment:
I think we can simplify this function. Now we run a single process reading live its output and making some action depending what was printed. `pyexpect` has function for it called `run` (which you can read more about it [here](https://pexpect.readthedocs.io/en/stable/api/pexpect.html#run-function)). Still the additional thread will be needed but the `_reader` function will be a lot simpler. Try to make PoC of it

Reply:
I've tested the approach with `pexpect.run`, it's much simpler, but there is some issue. Phoenixd does not close after completed loading or does not print some general message, that we could assert. That's why the killing utility used before is helpful (we can kill phoenixd process after successful passing copy/app cmd). Unfortunately `pexpect.run` does not return process, that we could kill (but only output and it's blocked until timeout occurred/some callback returned true).